### PR TITLE
[CIS-392] Safe sorting option for queries

### PR DESCRIPTION
**Pull request by mantergo**
**In this PR:**

- Add `cid` sorting key to `ChannelListSortingKey`
- Add safe sorting option to `ChannelListQuery`
- Add safe sorting option to `UserListQuery`

_`ChannelMemberListQuery` is left out cause there is no safe sorting option available._

### CODE REVIEW

File: Sources_v3/StreamChat/Query/Sorting/ChannelListSortingKey.swift [hunk 1]
```diff
@@ -16,6 +16,9 @@ public enum ChannelListSortingKey: String, SortingKey {
     case lastMessageAt
     /// Sort channels by number of members.
     case memberCount
+    /// Sort channels by `cid`.
+    /// **Note**: Avoid using this one cause it can be expensive from backend side.
```
Comment (VojtaStavik) [inline Sources_v3/StreamChat/Query/Sorting/ChannelListSortingKey.swift:20]: This sound quite harsh :) If we don't want users to use it (there's probably no reason why they even should), let's not make it public.

File: Sources_v3/StreamChat/Query/Sorting/ChannelListSortingKey.swift [hunk 2]
```diff
@@ -47,3 +47,15 @@ extension ChannelListSortingKey {
         .init(key: rawValue, ascending: isAscending)
     }
 }
+
+/// In case elements have same or `nil` target sorting values it's not possible to guarantee that order of elements will be the same
+/// all the time. So we need to additionally provide safe sorting option.
+extension Array where Element == Sorting<ChannelListSortingKey> {
+    func appendingSafeSortingIfNeeded() -> [Sorting<ChannelListSortingKey>] {
```
Comment (VojtaStavik) [inline Sources_v3/StreamChat/Query/Sorting/ChannelListSortingKey.swift:54]: Why not call it with a more descriptive name, for example `appendingCidSortingKey`? There's nothing like "safe sorting" :) Basically, what we're doing here is to make sure the sorting is always stable by appending a sort descriptor that's guaranteed to be unique.

**Review by VojtaStavik — APPROVED** [review]
Looks good, thank you 👍 Just some comments regarding naming.

File: Sources_v3/StreamChat/Query/Sorting/ChannelListSortingKey.swift [hunk 3]
```diff
@@ -43,3 +47,15 @@ extension ChannelListSortingKey {
...
+        guard !contains(where: { $0.key == .cid }), !isEmpty else {
+            return self
+        }
```
Comment (evsaev) [inline Sources_v3/StreamChat/Query/Sorting/ChannelListSortingKey.swift:57]: Do we really need `isEmpty == false` check? Shouldn't the `cid` sorting key be added if the array is empty?

File: Sources_v3/StreamChat/Query/Sorting/UserListSortingKey.swift [hunk 4]
```diff
@@ -40,3 +40,15 @@ extension UserListSortingKey {
...
     }
 }
+
+/// In case elements have same or `nil` target sorting values it's not possible to guarantee that order of elements will be the same
+/// all the time. So we need to additionally provide safe sorting option.
+extension Array where Element == Sorting<UserListSortingKey> {
+    func appendingSafeSortingIfNeeded() -> [Sorting<UserListSortingKey>] {
+        guard !contains(where: { $0.key == .id }), !isEmpty else {
+            return self
+        }
```
Comment (evsaev) [inline Sources_v3/StreamChat/Query/Sorting/UserListSortingKey.swift:50]: The same as https://github.com/GetStream/stream-chat-swift/pull/594/files#r525915508

File: Sources_v3/StreamChat/Query/Sorting/ChannelListSortingKey.swift (hunk 3 above)
Comment (b-onc) [reply Sources_v3/StreamChat/Query/Sorting/ChannelListSortingKey.swift:57]: AFAIK backend will sort them "safely" if client sends an empty sorting

Comment (mantergo) [reply Sources_v3/StreamChat/Query/Sorting/ChannelListSortingKey.swift:57]: Yep, @b-onc is right.
Backend will apply default sorting with secondary by `cid` in this case.

**Review by evsaev — APPROVED** [review]
👍

File: Sources_v3/StreamChat/Query/Sorting/ChannelListSortingKey.swift (hunk 1 above)
Comment (mantergo) [reply Sources_v3/StreamChat/Query/Sorting/ChannelListSortingKey.swift]: After discussions decided to go with updated comment.
See 34bee8a.

**Pull request merged**